lib/cog/pipeline.ex

```diff
@@ -0,0 +1,339 @@
+    prefix = Application.get_env(:cog, :command_prefix, "!")
+
+    text = text
+    |> String.replace(~r/^#{Regex.escape(prefix)}/, "") # Remove command prefix
+    |> String.replace(~r/“|”/, "\"")      # Replace OS X's smart quotes and dashes with ascii equivalent
+    |> String.replace(~r/‘|’/, "'")
+    |> String.replace(~r/—/, "--")
+    |> HtmlEntities.decode                # Decode html entities
+
+    # TODO: Fold this into decoding of the request initially?
+    %{request | text: text}
+  end
+
+  defp fetch_user(state) do
+    # TODO: This should happen when we validate the request
+    sender = state.request.sender.id
+    if ChatAdapter.is_chat_provider?(state.conn, state.request.provider) do
+      provider   = state.request.provider
+
+      user = UserQueries.for_chat_provider_user_id(sender, provider)
+      |> Repo.one
+
+      case user do
+        nil ->
+          {:error, :user_not_found}
+        user ->
+          {:ok, user}
+      end
+    else
+      case Repo.get_by(Cog.Models.User, username: sender) do
+        %Cog.Models.User{}=user ->
+          {:ok, user}
+        nil ->
+          {:error, :user_not_found}
+      end
+    end
+  end
+
+  defp parse(user, state) do
+    options = %ParserOptions{resolver: CommandResolver.command_resolver_fn(user)}
+    case Parser.scan_and_parse(state.request.text, options) do
+      {:ok, %Ast.Pipeline{}=pipeline} ->
+        case Destination.process(Ast.Pipeline.redirect_targets(pipeline),
+                                 state.request.sender,
+                                 state.request.room,
+                                 state.request.provider) do
+          {:ok, destinations} ->
+            {:ok, pipeline, destinations}
+          {:error, invalid} ->
+            {:error, {:redirect_error, invalid}}
+        end
+      {:error, msg} ->
+        {:error, {:parse_error, msg}}
+    end
+  end
+
+  defp start_pipeline(parsed, destinations, user, perms, state) do
+    case prepare_initial_context(state) do
+      {:ok, context} ->
+        case InitialContextSup.create([context: context, pipeline: self(), request_id: state.request.id]) do
+          {:ok, initial_context} ->
+            Process.monitor(initial_context)
+            try do
+              stages = parsed
+                       |> Enum.with_index
+                       |> Enum.reduce([initial_context], &create_stage!(&1, &2, user, perms, state))
+                       |> create_sinks!(parsed, destinations, user, state)
+              initialization_event(user, state)
+              InitialContext.unlock(initial_context)
+              {:reply, :ok, %{state | stages: stages}}
+            rescue
+              e in PrepareError ->
+                start_error_pipeline({:error, Exception.message(e)}, user, state)
+            end
+          error ->
+            start_error_pipeline(error, user, state)
+        end
+      error ->
+        start_error_pipeline(error, user, state)
+    end
+  end
```

Comment:
Guessing this was moved over form the old executor, but could use some `with` magic.